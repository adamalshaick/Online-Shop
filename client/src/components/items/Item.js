import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { deleteItem } from "../../actions/itemActions";
import { addItemToCart } from "../../actions/cartActions";
import { Link } from "react-router-dom";
import { ItemCard, Image, Buttons } from "../common/styles/StyledItem";
import { SecondaryHeader } from "../common/styles/Header";

export const Item = ({ item, currentUser, addItemToCart, deleteItem }) => {
  const onAdd = id => e => {
    e.preventDefault();
    addItemToCart(id);
  };

  const onDelete = id => e => {
    e.preventDefault();
    deleteItem(id);
  };

  return (
    <div className="col-md-6 col-lg-4 p-0 entry-2x">
      <div>
        <ItemCard className="text-center m-3">
          <Image src={item.itemImage} alt="" />
          <p>{item.name}</p>
          <SecondaryHeader>{item.price} $</SecondaryHeader>
          <span>
            {item.user === currentUser.id ? (
              <button
                id="deleteItem"
                onClick={onDelete(item._id)}
                className="btn btn-danger mr-1"
              >
                <i className="fas fa-times" />
              </button>
            ) : (
              <Buttons>
                {currentUser.cart.some(
                  cartItem => cartItem._id === item._id
                ) ? null : (
                  <>
                    <button
                      id="addToCart"
                      onClick={onAdd(item._id)}
                      className="btn btn-dark mb-2"
                    >
                      Add to your cart
                    </button>
                    <Link
                      to={`/profile/${item.user}`}
                      className="btn btn-dark mt-2"
                    >
                      Seller's profile
                    </Link>
                  </>
                )}
              </Buttons>
            )}
          </span>
        </ItemCard>
      </div>
    </div>
  );
};

Item.propTypes = {
  deleteItem: PropTypes.func.isRequired,
  addItemToCart: PropTypes.func.isRequired,
  item: PropTypes.object.isRequired,
  currentUser: PropTypes.object.isRequired
};

export default connect(
  null,
  { deleteItem, addItemToCart }
)(Item);
