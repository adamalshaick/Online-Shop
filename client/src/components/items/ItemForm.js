import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import TextAreaFieldGroup from "../common/TextAreaFieldGroup";

import { addItem } from "../../actions/itemActions";

class ItemForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      text: "",
      avatar: "",
      defaultPhoto: false,
      errors: {}
    };
  }

  componentWillReceiveProps(newProps) {
    if (newProps.errors) {
      this.setState({ errors: newProps.errors });
    }
  }

  onSubmit = e => {
    e.preventDefault();

    const { user } = this.props.auth;

    const newItem = {
      text: this.state.text,
      name: user.name,
      avatar: this.state.itemImage,
      itemImage: this.state.itemImage
    };

    this.props.addItem(newItem);
    this.setState({ text: "" });
    this.setState({ avatar: "" });
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { errors } = this.state;
    return (
      <div className="mb-3">
        {/* <div className="card card-info">
          <form>
            <ItemGroup placeholder="Add description" />
          </form>
        </div>
        <div className="card card-info">
          <div className="card-header bg-info text-white">
            Add an item for sale
          </div>
          <form onSubmit={this.onSubmit}>
            <input type="file" />

            <div className="form-group">
              <TextAreaFieldGroup
                className="form-control form-control-lg"
                placeholder="Add description"
                name="text"
                value={this.state.text}
                onChange={this.onChange}
                error={errors.text}
              />
            </div>
            <button type="submit" className="btn btn-info">
              Submit
            </button>
          </form>
        </div> */}
      </div>
    );
  }
}

ItemForm.propTypes = {
  addItem: PropTypes.func.isRequired,
  auth: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth,
  errors: state.errors
});

export default connect(
  mapStateToProps,
  { addItem }
)(ItemForm);
