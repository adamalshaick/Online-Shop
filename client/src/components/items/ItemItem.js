import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import classnames from "classnames";
import { Link } from "react-router-dom";
import { deleteItem } from "../../actions/itemActions";

class ItemItem extends Component {
  onDeleteClick(id) {
    this.props.deleteItem(id);
  }

  render() {
    const { item, auth, showActions } = this.props;

    return (
      <div className="col-md-4">
        <div style={{ height: "300px" }} className="card card-body mb-3 ">
          <div className="row">
            <div className="">
              <img
                style={{ maxWidth: "150px", maxHeight: "150px" }}
                className=" d-none d-md-block"
                src={`./../uploads/post_image/${item.itemImage}`}
                alt=""
              />

              <p>{item.price}</p>
            </div>
            <div className="">
              <p className="lead">{item.text}</p>
            </div>
            {showActions ? (
              <span>
                <Link to={`/item/${item._id}`} className="btn btn-info mr-1">
                  Items
                </Link>
                {item.user === auth.user.id ? (
                  <button type="button" className="btn btn-danger mr-1">
                    <i className="fas fa-times" />
                  </button>
                ) : null}
              </span>
            ) : null}
          </div>
        </div>
      </div>
    );
  }
}

ItemItem.defaultProps = {
  showActions: true
};

ItemItem.propTypes = {
  deleteItem: PropTypes.func.isRequired,
  item: PropTypes.object.isRequired,
  auth: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { deleteItem }
)(ItemItem);
