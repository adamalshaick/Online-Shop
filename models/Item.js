const mongoose = require("mongoose");
const Schema = mongoose.Schema;

//Create Schema
const ItemSchema = new Schema({
  user: {
    type: Schema.Types.ObjectId,
    ref: "users"
  },

  text: {
    type: String,
    required: true
  },

  price: {
    type: String,
    required: true
  },

  name: {
    type: String,
    required: true
  },

  itemImage: {
    type: String,
    required: true
  },

  category: {
    type: String,
    required: true
  },

  date: {
    type: Date,
    default: Date.now
  }
});

module.exports = Item = mongoose.model("item", ItemSchema);
