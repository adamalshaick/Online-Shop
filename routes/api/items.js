const express = require("express");
const router = express.Router();
const mongoose = require("mongoose");
const passport = require("passport");

// Item model
const Item = require("../../models/Item");
// Profile model
const Profile = require("../../models/Profile");

// Validation
const validateItemInput = require("../../validation/item");

// @route GET api/items/test
// @desc Tests items route
// @access Public
router.get("/test", (req, res) => res.json({ msg: "items works" }));

// @route GET api/items
// @desc Get items
// @access Public
router.get("/", (req, res) => {
  Item.find()
    .sort({ date: -1 })
    .then(items => res.json(items))
    .catch(err => res.status(404).json({ noitemsfound: "No items found" }));
});

// @route GET api/items/:id
// @desc Get item by id
// @access Public
router.get("/:id", (req, res) => {
  Item.findById(req.params.id)
    .then(item => res.json(item))
    .catch(err =>
      res.status(404).json({ noitemfound: "No item found with that ID" })
    );
});

// @route POST api/items
// @desc Create item
// @access Private
router.post(
  "/",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    const { errors, isValid } = validateItemInput(req.body);

    // Check Validation
    if (!isValid) {
      // If any errors, send 400 with errors object
      return res.status(400).json(errors);
    }
    const newItem = new Item({
      text: req.body.text,
      name: req.body.name,
      avatar: req.body.avatar,
      user: req.user.id
    });

    newItem.save().then(item => res.json(item));
  }
);

// @route DELETE api/items/:id
// @desc Delete item
// @access Private
router.delete(
  "/:id",
  passport.authenticate("jwt", { session: false }),
  (req, res) => {
    Profile.findOne({ user: req.user.id }).then(profile => {
      Item.findById(req.params.id)
        .then(item => {
          // Check for item owner
          if (item.user.toString() !== req.user.id) {
            return res
              .status(401)
              .json({ notauthorized: "User not authorized" });
          }

          //Delete
          item.remove().then(() => res.json({ succes: true }));
        })
        .catch(err => res.status(404).json({ itemnotfound: "No item found" }));
    });
  }
);

module.exports = router;
