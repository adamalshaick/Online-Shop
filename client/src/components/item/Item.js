import React, { Component } from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import ItemItem from "../items/ItemItem";
import { getItem } from "../../actions/itemActions";
import Link from "react-router-dom/Link";

export class Item extends Component {
  componentDidMount() {
    this.props.getItem(this.props.match.params.id);
  }

  render() {
    const { item, loading } = this.props.item;
    const { auth } = this.props;
    let itemContent;

    if (item === null || loading || Object.keys(item).length === 0) {
      itemContent = <p>Loading ...</p>;
    } else {
      itemContent = (
        <div>
          <ItemItem item={item} />
        </div>
      );
    }

    return (
      <div className="container">
        <div className="row">
          <div className="col-md-12">
            <Link to="/items" className="btn btn-light mb-3">
              Back to items
            </Link>
            {itemContent}
          </div>
        </div>
      </div>
    );
  }
}

Item.propTypes = {
  getItem: PropTypes.func.isRequired,
  item: PropTypes.object.isRequired
};

const mapStateToProps = state => ({
  item: state.item,
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { getItem }
)(Item);
