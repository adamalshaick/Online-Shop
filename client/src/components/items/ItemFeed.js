import React, { Component } from "react";
import PropTypes from "prop-types";
import Item from "./Item";

class ItemFeed extends Component {
  render() {
    const { items } = this.props;

    return (
      <>
        <h1 className="m-5">Items for sale</h1>
        <div className="row">
          {items.map(item => (
            <Item key={item._id} item={item} />
          ))}
        </div>
      </>
    );
  }
}

ItemFeed.propTypes = {
  items: PropTypes.array.isRequired
};

export default ItemFeed;
