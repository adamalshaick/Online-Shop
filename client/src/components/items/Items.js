import React, { Component } from "react";
import PropTypes from "prop-types";
import ItemFeed from "./ItemFeed";
import SelectListGroup from "../common/SelectListGroup";
import fetchItems from "../common/hoc/fetchItems";
import fetchCurrentUser from "../common/hoc/fetchCurrentUser";
import { SecondaryHeader } from "../common/styles/Header";
import {
  sortOptions,
  selectOptions,
  sort,
  select
} from "../../utils/browseItems";

class Items extends Component {
  constructor(props) {
    super(props);
    this.state = {
      sortBy: "",
      selectedCategory: ""
    };
  }

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { items } = this.props.item;
    const itemsArray = select(this.state.selectedCategory, items);
    sort(this.state.sortBy, itemsArray);

    return (
      <div className="container">
        <div className="row">
          <div className="col-md-12">
            <SecondaryHeader className="m-5 text-center">
              Items for sale
            </SecondaryHeader>
            <div className="row">
              <div className="col-md-6">
                <SelectListGroup
                  placeholder="* Category"
                  name="selectedCategory"
                  value={this.state.selectedCategory}
                  options={selectOptions}
                  onChange={this.onChange}
                />
              </div>
              <div className="col-md-6">
                <SelectListGroup
                  placeholder="* Sort by..."
                  name="sortBy"
                  value={this.state.sortBy}
                  options={sortOptions}
                  onChange={this.onChange}
                />
              </div>
            </div>
            <ItemFeed
              items={itemsArray}
              currentUser={this.props.user.currentUser}
            />
          </div>
        </div>
      </div>
    );
  }
}

Items.propTypes = {
  item: PropTypes.object.isRequired,
  cart: PropTypes.object.isRequired
};

export default fetchCurrentUser(fetchItems(Items));
