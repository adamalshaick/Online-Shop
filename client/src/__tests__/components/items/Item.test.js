import React from "react";
import { Item } from "../../../components/items/Item";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

const mockAddToCart = jest.fn();
const mockDelete = jest.fn();

describe("render component", () => {
  const Props = {
    item: {},
    currentUser: {}
  };

  const wrapper = shallow(
    <Item {...Props} deleteItem={mockDelete} addItemToCart={mockAddToCart} />
  );

  it("renders component", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("user owns the item", () => {
  const Props = {
    item: {
      user: "currentUser"
    },
    currentUser: {
      id: "currentUser"
    }
  };

  const wrapper = shallow(
    <Item {...Props} deleteItem={mockDelete} addItemToCart={mockAddToCart} />
  );

  it("displays delete button", () => {
    expect(wrapper.find("#deleteItem").length).toEqual(1);
  });

  it("calls mock delete function", () => {
    wrapper.find("#deleteItem").simulate("click", { preventDefault() {} });
    expect(mockDelete.mock.calls.length).toEqual(1);
  });

  it("doesn't display add to cart button", () => {
    expect(wrapper.find("#addToCart").length).toEqual(0);
  });
});

describe("user doesn't own the item", () => {
  describe("item is inside current users cart", () => {
    const PropsWithItemInCart = {
      item: {
        user: "anotherUser",
        _id: "itemInCart"
      },
      currentUser: {
        id: "currentUser",
        cart: [{ _id: "itemInCart" }, { _id: "anotherItem" }]
      }
    };

    const wrapper = shallow(
      <Item
        {...PropsWithItemInCart}
        deleteItem={mockDelete}
        addItemToCart={mockAddToCart}
      />
    );

    it("renders component", () => {
      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("doesn't display delete button", () => {
      expect(wrapper.find("#deleteItem").length).toEqual(0);
    });

    it("doesn't display add to cart button", () => {
      expect(wrapper.find("#addToCart").length).toEqual(0);
    });
  });

  describe("item is not inside current users cart", () => {
    const PropsWithoutItemInCart = {
      item: {
        user: "anotherUser",
        _id: "itemNotInCart"
      },
      currentUser: {
        id: "currentUser",
        cart: [{ _id: "itemInCart" }, { _id: "anotherItem" }]
      }
    };

    const wrapper = shallow(
      <Item
        {...PropsWithoutItemInCart}
        deleteItem={mockDelete}
        addItemToCart={mockAddToCart}
      />
    );

    it("doesn't display delete button", () => {
      expect(wrapper.find("#deleteItem").length).toEqual(0);
    });

    it("displays add to cart button", () => {
      expect(wrapper.find("#addToCart").length).toEqual(1);
    });

    it("calls mock add to cart function", () => {
      wrapper.find("#addToCart").simulate("click", { preventDefault() {} });
      expect(mockAddToCart.mock.calls.length).toEqual(1);
    });
  });
});
