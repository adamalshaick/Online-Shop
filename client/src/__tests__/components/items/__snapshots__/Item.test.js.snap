// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component renders component 1`] = `
<div
  className="col-md-6 col-lg-4 p-0 entry-2x"
>
  <div>
    <styled.div
      className="text-center m-3"
    >
      <styled.img
        alt=""
        src="../../uploads/post_image/placeholder.png"
      />
      <p />
      <styled.header>
         $
      </styled.header>
      <span>
        <button
          className="btn btn-danger mr-1"
          id="deleteItem"
          onClick={[Function]}
        >
          <i
            className="fas fa-times"
          />
        </button>
      </span>
    </styled.div>
  </div>
</div>
`;

exports[`user doesn't own the item item is inside current users cart renders component 1`] = `
<div
  className="col-md-6 col-lg-4 p-0 entry-2x"
>
  <div>
    <styled.div
      className="text-center m-3"
    >
      <styled.img
        alt=""
        src="../../uploads/post_image/placeholder.png"
      />
      <p />
      <styled.header>
         $
      </styled.header>
      <span>
        <styled.div />
      </span>
    </styled.div>
  </div>
</div>
`;
